### PR TITLE
feat: handle unsupported field names

### PR DESCRIPTION
**Pull request by y-lakhdar**
### Issue:
Whenever someone pushes documents containing unsupported field names, the entire push operation aborts because the automatic field creation is no longer possible.

### Proposition
Throw an error as soon as an invalid field is detected while creating a [Document](https://github.com/coveo/push-api-client.js/blob/main/src/document.ts#L52).

The user can provide a [Transformer](https://github.com/coveo/push-api-client.js/blob/main/src/validation/transformation/transformer.ts#L9) to normalize fields before creating the document.

The field should respect the following conditions: 
> Field names can only contain lowercase letters (a-z), numbers (0-9), and underscores. The field name must be at least one character long and must start with a lowercase letter.

### CODE REVIEW

File: src/source/push.ts [hunk 1]
```diff
@@ -284,13 +282,13 @@ export class PushSource {
     return axiosRequestHeaders(this.apikey);
   }
 
+  /**
```
Comment (y-lakhdar) [inline src/source/push.ts:285]: This is an error on my side. This method was never meant to be public at the first place.
If the user wants to detect/create missing fields, he should use the FieldAnalyser class

File: src/fieldAnalyser/fieldStore.ts [hunk 2]
```diff
@@ -11,11 +13,44 @@ export class FieldStore extends Map<string, FieldTypes> {
...
+  private formatIntoValidFieldName(fieldName: string): string {
+    fieldName = fieldName
+      // Extract alpha numeric only
+      .replace(/[^a-z0-9]/gi, '_')
+      // Remove any leading number or underscore
+      .replace(/^[0-9_]+/g, '')
+      // Remove consecutive underscores
+      .replace(/_{2,}/g, '_')
+      // Remove any trailing underscore
+      .replace(/_$/g, '');
```
Comment (y-lakhdar) [inline src/fieldAnalyser/fieldStore.ts:51]: It can probably be done in a single regex, but I find that way easier to read

**Comment by louis-bompart** [comment]
Detection and listing 👍 
Auto-fix 👎:

I think we should say the user what to do to sanitize his data, but sanitizing the field name in their stead is maybe a bad idea, because of the immutability of the name of a field.

Creating the missing one is OK because we're treating "valid input" and saving a call to the user, but here, we are doing some 'black magic' (yes, it's in the source code, but I don't expect the customer to go digging in it) to transform the fieldName.

So, to alleviate those points, while permitting fields name to be fixed on push, we should make it so the user is absolutely conscious of what is doing. To do so:
 - no auto-fix by default.
 - to auto-fix, the user has to provide a transformer string, i.e. `(input:string)=>string`
 - we can provide the one included in the PR as a static property, but the user would have to explicitly use it, i.e.:
 `new FieldStore(FieldStore.DefaultFieldNameFormater)` (i.e. the user explicitly feed the default formatter to the process)

**Comment by y-lakhdar** [comment]
> Detection and listing 👍 Auto-fix 👎:
> 
> I think we should say to the user what to do to sanitize his data, but sanitizing the field name in their stead is maybe a bad idea, because of the immutability of the name of a field.
> 
> Creating the missing ones is OK because we're treating "valid input" and saving a call to the user, but here, we are doing some 'black magic' (yes, it's in the source code, but I don't expect the customer to go digging in it) to transform the fieldName.
> 
> So, to alleviate those points, while permitting fields name to be fixed during a push, we should make it so the user is absolutely conscious of what they are doing. To do so:
> 
> * no auto-fix by default.
> * to auto-fix, the user has to provide a transformer string, i.e. `(input:string)=>string`
> * we can provide the one included in the PR as a static property, but the user would have to explicitly use it, i.e.:
>   `new FieldStore(FieldStore.DefaultFieldNameFormater)` (i.e. the user explicitly feed the default formatter to the process)

File: src/documentBuilder.ts [hunk 3]
```diff
@@ -2,7 +2,11 @@ import dayjs = require('dayjs');
 import {createHash} from 'crypto';
 import {CompressionType, Document, Metadata, MetadataValue} from './document';
 import {SecurityIdentityBuilder} from './securityIdentityBuilder';
-
+import {isFieldNameValid} from './fieldAnalyser/fieldUtils';
+import {UnsupportedFieldError} from './errors/fieldErrors';
+import {BuiltInTransformers} from './validation/transformation/builtInTransformers';
+// eslint-disable-next-line @typescript-eslint/no-unused-vars
```
Comment (louis-bompart) [inline src/documentBuilder.ts:8]: :suspect:

File: src/documentBuilder.ts [hunk 4]
```diff
@@ -127,11 +131,20 @@ export class DocumentBuilder {
...
+   * If not specified, no transformation will be applied to the metadata key.
+   *
+   * For a list of built-in transformers, use {@link BuiltInTransformers}.
    * @returns
    */
-  public withMetadataValue(key: string, value: MetadataValue) {
+  public withMetadataValue(
+    key: string,
+    value: MetadataValue,
+    keyTransformer = BuiltInTransformers.identity
```
Comment (louis-bompart) [inline src/documentBuilder.ts:145]: Prolly should type it :
```suggestion
    keyTransformer: Transformer = BuiltInTransformers.identity
```

File: src/documentBuilder.ts [hunk 5]
```diff
@@ -149,19 +162,43 @@ export class DocumentBuilder {
...
+   * @param {Transformer} [metadataKeyTransformer=BuiltInTransformers.identity] The {@link Transformer} to apply to all the document metadata keys.
+   * If not specified, no transformation will be applied to the metadata keys.
+   *
+   * For a list of built-in transformers, use {@link BuiltInTransformers}.
    * @returns
    */
-  public withMetadata(metadata: Metadata) {
+  public withMetadata(
+    metadata: Metadata,
+    metadataKeyTransformer = BuiltInTransformers.identity
```
Comment (louis-bompart) [inline src/documentBuilder.ts:184]:
```suggestion
    metadataKeyTransformer: Transformer = BuiltInTransformers.identity
```

File: src/errors/fieldErrors.ts [hunk 6]
```diff
@@ -37,3 +37,18 @@ export class InvalidPermanentId extends PushApiClientBaseError {
...
+  public name = 'Unsupported field Error';
+  public readonly unsupportedFields: string[];
+  public constructor(...unsupportedFields: string[]) {
+    super(`
+The following field names are invalid:
+${unsupportedFields.reduce(
+  (prev: string, curr: string) => (prev += `  • ${curr}\n`),
+  ''
+)}
+Field names can only contain lowercase letters (a-z), numbers (0-9), and underscores. The field name must be at least one character long and must start with a lowercase letter.
```
Comment (louis-bompart) [inline src/errors/fieldErrors.ts:50]: Stay consistent: 
```suggestion
Field names can only contain lowercase letters (a-z), numbers (0-9), and underscores. Field names must be at least one character long and must start with a lowercase letter.
```

File: src/fieldAnalyser/fieldUtils.ts [hunk 7]
```diff
@@ -36,3 +38,20 @@ export const createFields = async (
...
+  {
+    if (report.inconsistencies.size > 0) {
+      throw new FieldTypeInconsistencyError(report.inconsistencies);
+    }
+    await createFields(client, report.fields);
+  }
+};
+
+export const isFieldNameValid = (fieldName: string): boolean => {
+  const allowedChar = new RegExp('^[a-z]+[a-z0-9_]*$');
```
Comment (louis-bompart) [inline src/fieldAnalyser/fieldUtils.ts:55]: Put the RegExp outside the scope of the function, that will reduce the amount of call to 'new Regexp'

File: src/interfaces.ts [hunk 8]
```diff
@@ -1,5 +1,8 @@
 import type {AxiosResponse} from 'axios';
 import type {DocumentBuilder} from './documentBuilder';
+import type {Transformer} from './validation/transformation/transformer';
+// eslint-disable-next-line @typescript-eslint/no-unused-vars
```
Comment (louis-bompart) [inline src/interfaces.ts:4]: :suspect:

File: src/validation/parseFile.ts [hunk 9]
```diff
@@ -225,14 +231,14 @@ const processSecurityIdentities = (
...
+  options?: ParseDocumentOptions
 ) => {
+  const metadata: Metadata = {};
   Object.entries(caseInsensitiveDoc.documentRecord).forEach(([k, v]) => {
-    documentBuilder.withMetadataValue(
-      k,
-      v! as Extract<PrimitivesValues, MetadataValue>
-    );
+    metadata[k] = v! as Extract<PrimitivesValues, MetadataValue>;
   });
```
Comment (louis-bompart) [inline src/validation/parseFile.ts:240]: You don't use `k`, in that case, use `Object.values` instead.

**Review by louis-bompart — COMMENTED** [review]
LGTM overall, some small nit pic here and there

File: src/documentBuilder.ts (hunk 3 above)
Comment (y-lakhdar) [reply src/documentBuilder.ts:8]: For typedoc.
I need to import the type if I want to reference it in the documentation using links
image

File: src/interfaces.ts (hunk 8 above)
Comment (y-lakhdar) [reply src/interfaces.ts:4]: Reference in the documentation
image

File: src/validation/parseFile.ts (hunk 9 above)
Comment (y-lakhdar) [reply src/validation/parseFile.ts:240]: Not sure I get your point.
I need `k` to build my `Metadata` object right before calling the `withMetadata` method

Comment (louis-bompart) [reply src/validation/parseFile.ts:240]: nvm, I misread.

File: src/documentBuilder.ts (hunk 3 above)
Comment (louis-bompart) [reply src/documentBuilder.ts]: Yeah, ok. I tried some stuff but indeed that is a limitation with TS today.

To at least isolate this, do the import like this:
```ts
// Used for Documentation.
// eslint-disable-next-line @typescript-eslint/no-unused-vars
import type {BuiltInTransformers} from './validation/transformers/transformer';
```

File: src/interfaces.ts [hunk 10]
```diff
@@ -24,10 +29,24 @@ export interface BatchUpdateDocumentsOptions {
...
+   *
+   * For a list of built-in transformers, use {@link BuiltInTransformers}.
+   *
+   * @default BuiltInTransformers.identity
+   */
+  fieldNameTransformer?: Transformer;
+}
 export interface ConcurrentProcessing {
   /**
    * The maximum number of requests to send concurrently to the Coveo platform.
```
Comment (louis-bompart) [inline src/interfaces.ts:52]: While at it, let's add `@default 10 and remove the verbose 'the default value is set to'

**Review by louis-bompart — APPROVED** [review]

**Pull request merged**